build-scripts/version.py

```diff
@@ -0,0 +1,153 @@
+        version_h.write(VERSION_H)
+
+
+def write_VERSION_TXT():
+    url = "https://github.com/CleverRaven/Cataclysm-DDA"
+    if GITSHA:
+        url = f"{url}/commit/{GITSHA}"
+    timestamp = TIMESTAMP or datetime.datetime.now().strftime("%Y-%m-%d-%H%M")
+    try:
+        with open("VERSION.TXT", 'x') as VERSION_TXT:
+            text = str()
+            text += (f"build type: {ARTIFACT or 'Release'}\n"
+                     f"build number: {timestamp}\n"
+                     f"commit sha: {GITSHA or 'Unknown'}\n"
+                     f"commit url: {url}")
+            VERSION_TXT.write(text)
+    except FileExistsError:
+        log.debug('Skip writing VERSION.txt')
+
+
+# Not using argparse
+for arg in sys.argv[1:]:
+    arg = arg.split('=', 1)
+    log.debug(f"{arg}")
+    if len(arg) == 2:
+        globals()[arg[0]] = arg[1]
+
+ARTIFACT = ARTIFACT or os.environ.get('ARTIFACT', None) or 'Release'
+TIMESTAMP = TIMESTAMP or os.environ.get('TIMESTAMP', None)
+VERSION_STRING = VERSION or os.environ.get(
+    'VERSION', None) or os.environ.get('VERSION_STRING', None) or '0.I'
+GITSHA = None
+
+while not is_cwd_root():
+    # Assuming we started somewhere down, climb up to the source directory
+    os.chdir(Path.cwd().parent)
+
+# Checking for .git/ may not work because of external worktrees
+try:
+    git = subprocess.run(('git', 'rev-parse', '--is-inside-work-tree'),
+                         capture_output=True)
+except FileNotFoundError:  # `git` command is missing
+    write_version_h()
+    write_VERSION_TXT()
+    raise SystemExit
+
+if git.returncode != 0:
+    stdout = git.stdout.decode().strip()
+    if 'true' != stdout:
+        write_version_h()
+        write_VERSION_TXT()
+        raise SystemExit
+
+# Get the tag
+git = subprocess.run(('git', 'describe', '--tags', '--always',
+                     '--match', '[0-9A-Z]*.[0-9A-Z]*',
+                      '--match', 'cdda-experimental-*', '--exact-match'
+                      ),
+                     capture_output=True)
+GITVERSION = git.stdout.decode().strip()
+log.debug(f"{GITVERSION=}")
+
+# Get the short SHA1
+git = subprocess.run(('git', 'rev-parse', '--short', 'HEAD'),
+                     capture_output=True)
+GITSHA = git.stdout.decode().strip()
+log.debug(f"{GITSHA=}")
+
+# Check if there are changes in the worktree
+DIRTYFLAG = str()
+git = subprocess.run(('git', 'diff', '--numstat', '--exit-code',
+                      '-c', 'core.safecrlf=false'),
+                     capture_output=True)
+if git.returncode != 0:
+    stat = git.stdout.decode().strip()
+    # TODO filter lang/po
+    DIRTYFLAG = "-dirty"
+log.debug(f"{DIRTYFLAG=}")
+
+if GITVERSION:
+    VERSION_STRING = f"{GITVERSION} {GITSHA}{DIRTYFLAG}"
+else:
+    VERSION_STRING = f"{GITSHA}{DIRTYFLAG}"
+log.debug(f"{VERSION_STRING=}")
+
+OLDVERSION = read_version_h()
+
+if VERSION_STRING != OLDVERSION:
+    write_version_h()
+else:
+    log.debug("Skip writing src/version.h")
```

Comment:
Honestly, i would put this logic into `write_version_h()` itself. Pseudocode:
```py
def write_version_h(version):
    desired_content = ("//NOLINT(cata-header-guard)\n"
                     f'#define VERSION "{VERSION_STRING}"\n')
    current_content = None
    try:
        current_content = open("src/version.h", "r").read()
    except: pass
    if current_content != desired_content:
        open("src/version.h", "w").write(desired_content)
```

That's one less variable to worry about, and less code trying to parse the old version properly.